devito/ir/iet/scheduler.py

```diff
@@ -0,0 +1,158 @@
+from collections import OrderedDict
+
+import numpy as np
+
+from devito.cgen_utils import Allocator
+from devito.dimension import LoweredDimension
+from devito.ir.iet import (Expression, LocalExpression, Element, Iteration, List,
+                           UnboundedIndex, MetaCall, MapExpressions, Transformer,
+                           NestedTransformer, SubstituteExpression, iet_analyze,
+                           compose_nodes, filter_iterations, retrieve_iteration_tree)
+from devito.tools import filter_ordered, flatten
+from devito.types import Scalar
+
+__all__ = ['iet_build', 'iet_insert_C_decls']
+
+
+def iet_build(clusters, dtype):
+    """
+    Create an Iteration/Expression tree (IET) given an iterable of :class:`Cluster`s.
+    The nodes in the returned IET are decorated with properties deriving from
+    data dependence analysis.
+    """
+    # Clusters -> Iteration/Expression tree
+    iet = iet_make(clusters, dtype)
+
+    # Data dependency analysis. Properties are attached directly to nodes
+    iet = iet_analyze(iet)
+
+    # Substitute derived dimensions (e.g., t -> t0, t + 1 -> t1)
+    # This is postponed up to this point to ease /iet_analyze/'s life
+    subs = {}
+    for tree in retrieve_iteration_tree(iet):
+        uindices = flatten(i.uindices for i in tree)
+        subs.update({i.expr: LoweredDimension(name=i.index.name, origin=i.expr)
+                     for i in uindices})
+    iet = SubstituteExpression(subs).visit(iet)
+
+    return iet
+
+
+def iet_make(clusters, dtype):
+    """
+    Create an Iteration/Expression tree (IET) given an iterable of :class:`Cluster`s.
+
+    :param clusters: The iterable :class:`Cluster`s for which the IET is built.
+    :param dtype: The data type of the scalar expressions.
+    """
+    processed = []
+    schedule = OrderedDict()
+    for cluster in clusters:
+        if not cluster.ispace.empty:
+            root = None
+            intervals = cluster.ispace.intervals
+
+            # Can I reuse any of the previously scheduled Iterations ?
+            index = 0
+            for i0, i1 in zip(intervals, list(schedule)):
+                if i0 != i1 or i0.dim in clusters.atomics[cluster]:
+                    break
+                root = schedule[i1]
+                index += 1
+            needed = intervals[index:]
+
+            # Build Iterations, including any necessary unbounded index
+            iters = []
```

Comment:
note for reviewers (follows): In particular, this block of code has been pulled from ``ResolveIterationVariable``, refactored, and inserted here